File: src/go/svd/models/general_information.go
package models

import "time"

// GeneralInformation represents the "General" section of SVD.
type GeneralInformation struct {
	EventType             string        `json:"eventType" xml:"EventType" csv:"EventType"`
	OperationType         string        `json:"operationType" xml:"OperationType" csv:"OperationType"`
	OperationDescription  string        `json:"operationDescription" xml:"OperationDescription" csv:"OperationDescription"`
	PerformanceReportType string        `json:"performanceReportType" xml:"PerformanceReportType" csv:"PerformanceReportType"`
	ElapsedTime           time.Duration `json:"elapsedTime" xml:"ElapsedTime" csv:"ElapsedTime"`
	ShipLatitude          float64       `json:"shipLatitude" xml:"ShipLatitude" csv:"ShipLatitude"`
	ShipLongitude         float64       `json:"shipLongitude" xml:"ShipLongitude" csv:"ShipLongitude"`
	ShipReportingDate     time.Time     `json:"shipReportingDate" xml:"ShipReportingDate" csv:"ShipReportingDate"`
	ShipFlagState         string        `json:"shipFlagState" xml:"ShipFlagState" csv:"ShipFlagState"`
	ShipRegistryPortCode  string        `json:"shipRegistryPortCode" xml:"ShipRegistryPortCode" csv:"ShipRegistryPortCode"`
	ShipRegistryPortName  string        `json:"shipRegistryPortName" xml:"ShipRegistryPortName" csv:"ShipRegistryPortName"`
	ShipName              string        `json:"shipName" xml:"ShipName" csv:"ShipName"`
	Imo                   string        `json:"imo" xml:"Imo" csv:"Imo"`
	Mmsi                  string        `json:"mmsi" xml:"Mmsi" csv:"Mmsi"`
	ShipType              string        `json:"shipType" xml:"ShipType" csv:"ShipType"`
	ShipTypeMarpolAnnexVi string        `json:"shipTypeMarpolAnnexVi" xml:"ShipTypeMarpolAnnexVi" csv:"ShipTypeMarpolAnnexVi"`
	NumberOfPassengers    int           `json:"numberOfPassengers" xml:"NumberOfPassengers" csv:"NumberOfPassengers"`
	NumberOfCrew          int           `json:"numberOfCrew" xml:"NumberOfCrew" csv:"NumberOfCrew"`
	VoyageNumber          string        `json:"voyageNumber" xml:"VoyageNumber" csv:"VoyageNumber"`
	VoyageRemarks         string        `json:"voyageRemarks" xml:"VoyageRemarks" csv:"VoyageRemarks"`
	VoyageLegIdentifier   string        `json:"voyageLegIdentifier" xml:"VoyageLegIdentifier" csv:"VoyageLegIdentifier"`
	VoyageLegRemarks      string        `json:"voyageLegRemarks" xml:"VoyageLegRemarks" csv:"VoyageLegRemarks"`
}

File: src/elixir/svd/lib/svd/models/general_information.ex
defmodule SVD.Models.GeneralInformation do
  @moduledoc false

  @fields [
    :event_type,
    :operation_type,
    :operation_description,
    :performance_report_type,
    :elapsed_time,
    :ship_latitude,
    :ship_longitude,
    :ship_reporting_date,
    :ship_flag_state,
    :ship_registry_port_code,
    :ship_registry_port_name,
    :ship_name,
    :imo,
    :call_sign,
    :mmsi,
    :ship_type,
    :ship_type_marpol_annex_vi,
    :number_of_passengers,
    :number_of_crew,
    :voyage_number,
    :voyage_remarks,
    :voyage_leg_identifier,
    :voyage_leg_remarks,
    :deadweight,
    :gross_tonnage,
    :net_tonnage,
    :ship_owner,
    :ship_manager,
    :charterer,
    :length_overall,
    :beam,
    :depth
  ]

  @derive {Jason.Encoder, only: @fields}
  defstruct @fields

  @type t :: %__MODULE__{}

  @spec fields() :: [atom()]
  def fields, do: @fields
end
